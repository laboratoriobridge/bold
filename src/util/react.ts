import { MutableRefObject, Ref } from 'react'

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}
