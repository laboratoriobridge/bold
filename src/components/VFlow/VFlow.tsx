import React, { CSSProperties } from 'react'

import { ExternalStyles, Theme, useStyles } from '../../styles'
import { Omit } from '../../util'

export interface VFlowProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'style'> {
  vSpacing?: number
  style?: ExternalStyles
}

export function VFlow(props: VFlowProps) {
  const { vSpacing, style, children, ...rest } = props
  const { classes, css } = useStyles(createStyles, props)

  return (
    <div className={css(classes.container, style)} {...rest}>
      {React.Children.map(children, (child) => child && <div className={classes.childWrapper}>{child}</div>)}
    </div>
  )
}

VFlow.defaultProps = {
  vSpacing: 1,
} as VFlowProps

const createStyles = (theme: Theme, { vSpacing }: VFlowProps) => ({
  container: {
    display: 'flex',
    flexDirection: 'column',
  } as CSSProperties,
  childWrapper: {
    display: 'inline-block',
    marginBottom: `${vSpacing / 2}rem`,
    marginTop: `${vSpacing / 2}rem`,

    ':first-of-type': {
      marginTop: 0,
    },
    ':last-of-type': {
      marginBottom: 0,
    },
    ':empty': {
      display: 'none',
    },
  } as CSSProperties,
})
