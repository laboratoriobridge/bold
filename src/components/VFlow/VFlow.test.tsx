import { render } from '@testing-library/react'
import React from 'react'

import { VFlow } from './VFlow'

it('should render correctly', () => {
  const { container } = render(
    <VFlow>
      <span>1</span>
      <span>2</span>
      <span>3</span>
      456
    </VFlow>
  )
  expect(container).toMatchSnapshot()
})

it('should accept "style" prop', () => {
  const { container } = render(
    <VFlow style={{ color: 'red' }}>
      <span>1</span>
      <span>2</span>
      <span>3</span>
    </VFlow>
  )
  expect(container).toMatchSnapshot()
})

it('should render wrappers only for valid children', () => {
  const { container } = render(
    <VFlow>
      {null}
      {undefined}
      <span>testing vflow</span>
      {true}
      {false}
      <span>testing vflow</span>
    </VFlow>
  )
  expect(container.querySelectorAll('div')[0].childNodes.length).toEqual(2)
  expect(container).toMatchSnapshot()
})
