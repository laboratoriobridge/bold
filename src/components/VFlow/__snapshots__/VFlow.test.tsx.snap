// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept "style" prop 1`] = `
.emotion-0 {
  display: inline-block;
  margin-bottom: 0.5rem;
  margin-top: 0.5rem;
}

.emotion-0:first-of-type {
  margin-top: 0;
}

.emotion-0:last-of-type {
  margin-bottom: 0;
}

.emotion-0:empty {
  display: none;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: red;
}

<div>
  <div
    class="emotion-3"
  >
    <div
      class="emotion-0"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="emotion-0"
    >
      <span>
        2
      </span>
    </div>
    <div
      class="emotion-0"
    >
      <span>
        3
      </span>
    </div>
  </div>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  display: inline-block;
  margin-bottom: 0.5rem;
  margin-top: 0.5rem;
}

.emotion-0:first-of-type {
  margin-top: 0;
}

.emotion-0:last-of-type {
  margin-bottom: 0;
}

.emotion-0:empty {
  display: none;
}

<div>
  <div
    class="emotion-4"
  >
    <div
      class="emotion-0"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="emotion-0"
    >
      <span>
        2
      </span>
    </div>
    <div
      class="emotion-0"
    >
      <span>
        3
      </span>
    </div>
    <div
      class="emotion-0"
    >
      456
    </div>
  </div>
</div>
`;

exports[`should render wrappers only for valid children 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  display: inline-block;
  margin-bottom: 0.5rem;
  margin-top: 0.5rem;
}

.emotion-0:first-of-type {
  margin-top: 0;
}

.emotion-0:last-of-type {
  margin-bottom: 0;
}

.emotion-0:empty {
  display: none;
}

<div>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-0"
    >
      <span>
        testing vflow
      </span>
    </div>
    <div
      class="emotion-0"
    >
      <span>
        testing vflow
      </span>
    </div>
  </div>
</div>
`;
